src/main/java/pl/project13/core/ProcessHandler.java

```diff
@@ -0,0 +1,160 @@
+  private final Future<Void> stdoutFuture;
+  private final Future<String> stderrFuture;
+
+  private String stderrOutput = null;
+
+  /**
+   * @param process the process which should be handled
+   * @param stdoutLineConsumer called asynchronously with the lines read from stdout. The consumer
+   *        must either be thread-safe, or the result it is building must only be used after
+   *        {@link #exitValue(long, TimeUnit)} has returned without throwing an exception. The
+   *        consumer must not block, otherwise this could prevent the process from writing any
+   *        output, causing it to get stuck.
+   */
+  public ProcessHandler(Process process, Consumer<String> stdoutLineConsumer) {
+    this.process = Objects.requireNonNull(process);
+    Objects.requireNonNull(stdoutLineConsumer);
+
+    // 2 threads, one for stdout, one for stderr
+    // The process output is consumed concurrently by separate threads because otherwise the process
+    // could get stuck if the output is not consumed and the output buffer is full
+    ThreadFactory threadFactory = Executors.defaultThreadFactory();
+    outputReaderExecutor = Executors.newFixedThreadPool(2, runnable -> {
+      Thread t = threadFactory.newThread(runnable);
+      // Don't prevent JVM exit
+      t.setDaemon(true);
+      return t;
+    });
+
+    String processInfo;
+    try {
+      processInfo = this.process.info().command().orElse("?") + " [" + this.process.pid() + "]";
+    } catch (UnsupportedOperationException e) {
+      processInfo = "<unknown-process>";
+    }
+    stdoutFuture =
+        outputReaderExecutor.submit(new ProcessOutputReader<>("stdout reader (" + processInfo + ")",
+            this.process.getInputStream(), stdoutLineConsumer,
+            // Don't create a 'result', `stdoutLineConsumer` will do that itself if needed
+            () -> null));
+
+    StringBuilder stderrBuilder = new StringBuilder();
+    stderrFuture =
+        outputReaderExecutor.submit(new ProcessOutputReader<>("stderr reader (" + processInfo + ")",
+            this.process.getErrorStream(), line -> stderrBuilder.append(line).append('\n'),
+            stderrBuilder::toString));
+  }
+
+  /**
+   * Waits for the process to finish and returns the exit value.
+   *
+   * @throws TimeoutException if waiting for the process to finish times out
+   */
+  public int exitValue(long timeout, TimeUnit timeUnit) throws InterruptedException, TimeoutException, ExecutionException {
+    boolean finished = process.waitFor(timeout, timeUnit);
+    if (finished) {
+
+      outputReaderExecutor.shutdown();
+      try {
+        stdoutFuture.get();
+      } catch (ExecutionException e) {
+        throw new ExecutionException("Failed waiting for stdout", e.getCause());
+      }
+      try {
+        stderrOutput = stderrFuture.get();
+      } catch (ExecutionException e) {
+        throw new ExecutionException("Failed waiting for stderr", e.getCause());
+      }
```

Comment:
Not completely sure about this unwrapping with `e.getCause()` and then rewrapping as `ExecutionException`. Could also directly propagate `e`, but then you would have to inspect the stack trace to understand whether the exception occurred for the stdout or stderr thread.